### DISCUSSION TITLE
2024-11-01 - REST API Design - Chewing the Fat 📈

### DISCUSSION THREAD

**Discussion by sethdaily** [SSWConsulting/SSW.Rules.Content#9512] [Chewing the Fat]
Hey programmers,
Today we're talking about API design - [Do you know the key principles of REST API Design?](https://www.ssw.com.au/rules/rest-api-design)

* What’s a good example of a public API you've worked with, and why?
* Any other API design tips you’d recommend?

**Comment by ozdemir-mehmet** [comment]
First post
The ABS API was pretty good when I last used it, Give options around bulk data egress/ingress. Don't rely on 10 row pagination on a 1,000,000 row dataset

**Comment by TobyChurches** [comment]
Great topic and very relevant! API design (particularly Rest API's) is super important, making sure that the developer can easily, efficiently and securely access the required data. As for good API's, none really come to mind - Its the bad ones that are memorable. I recently had to work with the Google maps API and hated it. It seemed to be too opinionated for our use case.

**Comment by calumjs** [comment]
>What’s a good example of a public API you've worked with, and why?

The Dynamics API is cool

>Any other API design tips you’d recommend?

Return useful error messages with tips on how to fix - this makes it easier to chatgpt it

**Comment by drwharris** [comment]
> What’s a good example of a public API you've worked with, and why?

I loved working with the Stripe API. Well thought out, well structured and well documented. 

> Any other API design tips you’d recommend?

Try documenting your API before releasing it, its a true sign of how easy your API will be to implement.

**Comment by leahy268** [comment]
Never built an API.  Have gathered data from a few bad and good ones though.  The worst I think that I've found so far was the YouTube API.  Took me hours to try and understand how to get data from it.  Could have been the poor documentation though.

**Comment by jeoffreyfischer** [comment]
> What’s a good example of a public API you've worked with, and why?

None - I only worked with private ones

> Any other API design tips you’d recommend?

Return useful error messages for debugging

**Comment by joshbermanssw** [comment]
> What’s a good example of a public API you've worked with, and why?

n/a - only worked with private 

> Any other API design tips you’d recommend?

Correct error handling, validation and error returning are key!

**Comment by StefStar** [comment]
I'm not a developer however, I learnt that API's are the backbone of a lot of developer work and not having clear names can easily cause headaches down the line.

**Comment by steven0x51** [comment]
> Any other API design tips you’d recommend?

Error Handling - Use a consistent format, including status codes and error messages gives user better experience and easier to debug

**Comment by Lewkans** [comment]
> What’s a good example of a public API you've worked with, and why?

None - haven't really dealt with public APIs much

> Any other API design tips you’d recommend?

Returning relevant and detailed error messages instead of generic error messages for easier debugging in the future.

**Comment by Ben0189** [comment]
cant think of a good one atm, but Xero API is not great

**Comment by RobJT-wq** [comment]
I've only used the Microsoft Graph API to automate some SysAdmin tasks, It's pretty easy to work with for my needs.

**Comment by Calinator444** [comment]
> What’s a good example of a public API you've worked with, and why?

I've worked with the ChatGPT completions API. It provides a really robust set of features:
- It actually has a more robust feature set than the ChatGPT web client (e.g. providing a "temperature" field to alter the creativeness of the responses).
- It's got an officially supported Python library that conceals the complexity of the HTTP requests

> Any other API design tips you’d recommend?

Use path parameters for retrieving resources and query params for filtering

e.g. 

use this **api/posts/1**

not this **api/posts?id=1**

**Comment by danielmackay** [comment]
> What’s a good example of a public API you've worked with, and why?

Binance - https://www.binance.com/en-NZ/binance-api

This has great documentation that covers both rest and websocket APIs, and also higher-level general documentation.

> Any other API design tips you’d recommend?

No, I put everything I could think of in the rule. 😄

**Comment by ricksu978** [comment]
> What’s a good example of a public API you've worked with, and why?

Clear documentation!
- Stripe: https://docs.stripe.com/api
- Shopify: https://shopify.dev/docs/api

> Any other API design tips you’d recommend?

# Use colon for custom verbs
https://google.aip.dev/136

**Comment by brookjeynes-ssw** [comment]
Architect to the size of your system. There's no need to over-complicate a simple REST app with architecture designs that don't matter in the end if the system size permits for that approach.

**Comment by jimmy1992abc** [comment]
> What’s a good example of a public API you've worked with, and why?

I have user experience with Xero and Microsoft Speech Studio API. I am not a developer and may not understand all the details, but I find both of them use GUID (which can be generated by them or your license key) at the URL to make it unique. Later, their servers may direct users to the real API.

> Any other API design tips you’d recommend?

Not a developer, no idea.

**Comment by chrisschultzssw** [comment]
I've never built an API, but when connecting to an API the most useful thing to me has been good documentation.

**Comment by ncn-ssw** [comment]
> What’s a good example of a public API you've worked with, and why?

GitHub GraphQL API - because the documentation and error messages were useful
 
> Any other API design tips you’d recommend?

Error messages - When a call fails, ensure that a clear description of why is returned. If appropriate, a link to the appropriate page in the API documentation should be included.

**Comment by isaaclombardssw** [comment]
Already been said but the documentation makes heaps of difference, bonus if I can concisely see what routes are available to me with some info like https://worldtimeapi.org/api.

**Comment by sam-s-maher** [comment]
> What’s a good example of a public API you've worked with, and why?

Easier to remember bad ones than good ones, and for me it's always down to the following...
 
> Any other API design tips you’d recommend?

Clear documenation! Especially if it's public. For private API's, onboarding is also easier with good documentation. On that topic, has anyone ever used Postman to manage API documentation within a team?

**Comment by JackDevAU** [comment]
Sometime we need to remember the scope of the project when going into architecting a system. I've seen a lot of Apis that are overcooked just because they can be, but not because they should be.

**Comment by amankumarrr** [comment]
Returning accurate error messages is essential for effective debugging, both for consumers of the API and for developers on the backend. Clear error messages provide critical context, making it easier to identify, troubleshoot, and resolve issues efficiently on both sides.

**Comment by tkapa** [comment]
Scryfall API is quite good.
It provides a very well thought out and thorough API for Magic the Gathering cards.

As nick pointed out, nice Errors are important to creating a nice API

**Comment by awaltos** [comment]
> What’s a good example of a public API you've worked with, and why?

Only public one i used is the pokemon one www.pokeapi.co
It is a bit complex but there are good docs and preview/testing features on the site.

> Any other API design tips you’d recommend?

Having solid error handling and meaningful error responses should be a priority

**Comment by MattParkerDev** [comment]
> What’s a good example of a public API you've worked with, and why?

I don't think I have any good examples, only bad examples (of private APIs 😅)

> Any other API design tips you’d recommend?

don't overcook it, keep it simple, and logically group and nest your endpoints

**Comment by Tanya-Leahy** [comment]
Not a dev and have had limited experience with APIs. The only one that comes to mind is the YouTube API which I worked on with Warwick and it was awful.

**Comment by bettybondoc** [comment]
I've answered "NA - Not a Developer" in the form cause I haven't touched APIs since joining SSW but I remember having fun with Google Maps and Spotify APIs in past projects.

**Comment by matt-goldman-ssw** [comment]
> What’s a good example of a public API you've worked with, and why?

NASA 


> Any other API design tips you’d recommend?

Don't use API keys, use client credentials

**Comment by PennyWalker** [comment]
This was a great video and excellent for the devs, but I haven't got enough background knowledge to answer these questions :)

**Comment by stephanfakossw** [comment]
I've learned that I don't know anything about this, so I can't contribute to the discussion...I trust it's an excellent topic for the devs

**Comment by AttackOnMorty** [comment]
> What’s a good example of a public API you've worked with, and why?

Spotify - It has an online playground.
https://developer.spotify.com/documentation/web-api

> Any other API design tips you’d recommend?

Documentation

**Comment by bahjatalaadelssw** [comment]
> What’s a good example of a public API you've worked with, and why?
I'm interested in using AI to work with an API
> Any other API design tips you’d recommend?
Universal naming conventions

**Comment by SamBWagner** [comment]
> What’s a good example of a public API you've worked with, and why?

- ScryFall! It's awesome 
- GitHub

> Any other API design tips you’d recommend?

keep your API as flat as you can. No more than 2 layers if you can manage it i.e. /{something}/{something}

**Comment by jimmidier** [comment]
> What’s a good example of a public API you've worked with, and why?

None. Haven't worked on public API yet.

> Any other API design tips you’d recommend?

Consistent response format with response code / error code

**Comment by thom-wang** [comment]
> What’s a good example of a public API you've worked with, and why?

I think GitHub's REST API is quite good to work with because it has rich documentation and is beautifully designed meaning every standard a good API should have.

> Any other API design tips you’d recommend?

Use GET less and replace it with POST as many as possible, cause it will help to reduce CORS attack.

**Comment by babakamyljanovssw** [comment]
> What’s a good example of a public API you've worked with, and why?

GitHub API, because it is well-documented with useful ready commands for each endpoints and tools to test them out

> Any other API design tips you’d recommend?

Standard status code - using status codes that matches outcome: 200 OK, 201 Created, 400 Bad Request, 404 Not Found, 500 Internal Server Error

Error handling - helpful error messages can help to understand what went wrong

**Comment by jerryluossw** [comment]
> Any other API design tips you’d recommend?

- A simple consistent response format
- Use PUT and DELETE cautiously, as many organizations’ security policies may block them by default

**Comment by vladislav-kir** [comment]
> What’s a good example of a public API you've worked with, and why?

Modern eCommerce Gateways such as Stripe and Square, Microsoft Graph API, Dynamics 365 API - they have follow most of the tips and have a great documentation. Although, for Dynamics 365 API query string parameters can get very long and messy....

> Any other API design tips you’d recommend?

Good error handling - clearly explain what the problem is with the request.

**Comment by JeanThirion** [comment]
> What’s a good example of a public API you've worked with, and why?

https://radarr.video/docs/api/
(same for Sonarr and all the rr's)

Very clear documentation, accessible logs, ... 
MSGraph is pretty good too to be honest !

**Comment by PothieuG** [comment]
>What’s a good example of a public API you've worked with, and why?

Github API is very good and well documented!

**Comment by ozair-ashfaque** [comment]
> What’s a good example of a public API you've worked with, and why?

None - I only worked with the private ones.

> Any other API design tips you’d recommend?

Error handling - Provide clear, structured error responses with relevant HTTP status codes and helpful messages to guide clients in resolving issues.

**Comment by tino-liu** [comment]
> What’s a good example of a public API you've worked with, and why?

None - Only worked on private ones

> Any other API design tips you’d recommend?

Recent work taught me that clear error handling and logging them is crucial

**Comment by Aibono1225** [comment]
> What’s a good example of a public API you've worked with, and why?

None - only worked on private APIs

> Any other API design tips you’d recommend?

Standard status code
Error handling + logging
